python/phase-1.md

```diff
@@ -0,0 +1,380 @@
+
+* Phase 1 - Initial project, activity worker, and client
+* Phase 2 - Workflow worker, local activities, async activity completion, etc
+* Phase 3 - Deterministic workflow sandbox, replayer, test suite (some parts may be combined with phase 2)
+
+This proposal is for phase 1. It is intentionally loosely defined and all aspects are subject to discussion/change.
+
+## Overview
+
+The Python SDK will be backed by [sdk-core](https://github.com/temporalio/sdk-core), utilizing a bridge based on
+[PyO3](https://github.com/PyO3/PyO3) similar to `sdk-typescript`'s `neon`-based bridge.
+
+The high-level goals for the Python SDK are:
+
+* Be as simple as possible from a caller POV
+* Be as lightweight as possible wrt dependencies
+* Be reasonably flexible with reasonable defaults
+* Support only Python >= 3.7
+* Be friendly to REPL/interpreter use cases
+
+The stack/approach looks like:
+
+* Python >= 3.7 (so may need `typing_extension` for things like `ParamSpec`)
+* Poetry packaging, black formatting
+  * Pipenv originally considered
+* PyO3 + sdk-core
+* asyncio direct (i.e. no need for anyio-style wrapper)
+* Type annotations everywhere
+* Decorator-based activity and workflow definitions
+
+An example of a client (intentionally missing _a lot_ of features):
+
+```python
+import temporalio
+
+def main():
+  client = temporalio.Client("localhost:7233", namespace="MyNamespace")
+  result = await client.execute_workflow("MyWorkflow", "arg1", id="someid", task_queue="somequeue")
+  print("Result", result)
+```
+
+An example of an activity and registration (also missing a lot of features):
+
+```python
+import temporalio
+
+def say_hello(name: str) -> str:
+  return f"Hello, {name}!"
+
+def main(client: temporalio.Client):
+  worker = temporalio.Worker(client, task_queue="somequeue")
+  worker.register_activity(say_hello)
+  await worker.run()
+```
+
+## API Definitions
+
+A lot of loosely-defined API definitions are below. They are intentionally lacking details.
+
+High-level notes:
+
+* For demonstration/simplicity here we are leaving off a lot of proper typings (e.g. `Optional`) that the real version
+  will have
+* We leave off a lot of options on calls for this high-level proposal
+* We are not using proper function documentation, but rather inline comments for easier discussion
+* **NOTE: These will have proper docs and types when developed!**
+
+### Client
+
+Exported from `temporalio` (but probably in `temporalio/client/client.py`, exported from `temporalio.client` then
+exported from `temporalio`):
+
+```python
+class Client:
+    def __init__(
+        self,
+        addr: str,
+        *,
+        namespace: str = "default",
+        default_task_queue: str = None,
+        identity: str = None
+        # ... additional options omitted for brevity
+    ) -> None:
+        # Actually dials gRPC here
+        pass
+
+    # Raw gRPC service
+    @property
+    def service(self) -> workflowservice.WorkflowService:
```

Comment:
I'd expect the `service` to be provided to via constructor so it can be shared by multiple clients.

Reply:
That adds an unnecessary second step. Creating a client and using it is very reasonable/simple.

However since I am switching to a static `create` method for `await` purposes, so I'll have an overload for those that really want to manually create a gRPC client.

Reply:
👍